app/models/voxelytics/VoxelyticsService.scala

```diff
@@ -118,10 +150,41 @@ class VoxelyticsService @Inject()(voxelyticsDAO: VoxelyticsDAO)(implicit ec: Exe
       runUserId <- voxelyticsDAO.getUserIdForRunOpt(runName, user._organization)
     } yield bool2Fox(user.isAdmin || runUserId.forall(_ == user._id))
 
-  def runPublicWrites(run: RunEntry, tasks: List[TaskRunEntry]): JsObject =
-    Json.toJson(run).as[JsObject] ++ Json.obj(
-      "tasks" -> tasks.map(Json.toJson(_))
-    )
+  def taskRunsPublicWrites(combinedTaskRuns: List[CombinedTaskRunEntry], taskRuns: List[TaskRunEntry]): JsArray = {
+    val groupedTaskRuns = taskRuns.groupBy(_.taskName)
+    JsArray(
+      groupedTaskRuns
+        .map(group => {
+          val sortedTaskRuns = group._2.sortBy(_.beginTime).reverse
+          val combinedTaskRun = combinedTaskRuns.find(_.taskName == group._1)
+          Json.obj(
+            "taskName" -> group._1,
+            "state" -> sortedTaskRuns
+              .map(_.state)
+              .find(s => s != VoxelyticsRunState.SKIPPED && s != VoxelyticsRunState.PENDING)
+              .orElse(sortedTaskRuns.headOption.map(_.state))
+              .getOrElse(VoxelyticsRunState.SKIPPED)
+              .asInstanceOf[VoxelyticsRunState],
+            "beginTime" -> Try(sortedTaskRuns.flatMap(_.beginTime).min).toOption,
+            "endTime" -> Try(sortedTaskRuns.flatMap(_.endTime).max).toOption,
+            "currentExecutionId" -> combinedTaskRun.flatMap(_.currentExecutionId),
+            "chunks" -> combinedTaskRun
+              .map(_.chunks)
+              .getOrElse(ChunkStatistics(0, 0, 0, 0, 0))
+              .asInstanceOf[ChunkStatistics],
```

Comment:
Again it is not obvious to me why a cast is needed. Don’t the items have this type already?